components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,240 @@ package xid
-		if unwrapped, ok := o.Unwrap().(*dmesg.Component); ok {
-			dmesgComponent = unwrapped
-		} else {
-			return nil, fmt.Errorf("expected *dmesg.Component, got %T", dmesgC)
-		}
-	}
-	dmesgTailResults, err := dmesgComponent.TailScan()
+func (c *XIDComponent) Start() error {
+	watcher, err := pkg_dmesg.NewWatcher()
 	if err != nil {
-		return nil, err
+		log.Logger.Errorw("failed to create dmesg watcher", "error", err)
+		return nil
 	}
-
-	o := &Output{}
-	for _, logItem := range dmesgTailResults.TailScanMatched {
-		if logItem.Error != nil {
+	for {
+		backoffTime := 1 * time.Second
+		osComponent, err := components.GetComponent(os_id.Name)
+		if err != nil {
+			log.Logger.Errorw("failed to get os component", "error", err)
+			time.Sleep(backoffTime)
 			continue
 		}
-		if logItem.Matched == nil {
+		osEvents, err := osComponent.Events(c.rootCtx, time.Now().Add(-3*24*time.Hour))
+		if err != nil {
+			log.Logger.Errorw("failed to get os states", "error", err)
+			time.Sleep(backoffTime)
 			continue
 		}
-		if logItem.Matched.Name != dmesg.EventNvidiaNVRMXid {
+		if len(osEvents) < 1 {
+			log.Logger.Debugw("no os states found")
+			time.Sleep(backoffTime)
 			continue
 		}
-
-		ev, err := nvidia_query_xid.ParseDmesgLogLine(logItem.Time, logItem.Line)
+		localEvents, err := c.store.Get(c.rootCtx, time.Time{})
 		if err != nil {
-			return nil, err
+			log.Logger.Errorw("failed to get all events", "error", err)
+			time.Sleep(backoffTime)
+			continue
 		}
-		o.DmesgErrors = append(o.DmesgErrors, ev)
-	}
-
-	last, err := c.poller.LastSuccess()
-
-	// no data yet from realtime xid poller
-	// just return whatever we got from dmesg
-	if err == query.ErrNoData {
-		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", nvidia_component_error_xid_id.Name)
-		return o, nil
+		events := mergeEvents(osEvents, localEvents)
+		c.Lock()
+		c.currState = EvolveHealthyState(events)
+		c.Unlock()
+		break
 	}
-
-	// something went wrong in the poller
-	// just return an error to surface the issue
-	if err != nil {
-		return nil, err
-	}
-
-	ev, ok := last.Output.(*nvidia_query_nvml.XidEvent)
-	if !ok { // shoild never happen
-		return nil, fmt.Errorf("invalid output type: %T, expected nvidia_query_nvml.XidEvent", last.Output)
-	}
-	if ev != nil && ev.Xid > 0 {
-		o.NVMLXidEvent = ev
-
-		if lerr := c.poller.LastError(); lerr != nil {
-			log.Logger.Warnw("last query failed -- returning cached, possibly stale data", "error", lerr)
+	go func() {
+		ticker := time.NewTicker(DefaultStateUpdatePeriod)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.rootCtx.Done():
+				return
+			case <-ticker.C:
+				osComponent, err := components.GetComponent(os_id.Name)
+				if err != nil {
+					log.Logger.Errorw("failed to get os component", "error", err)
+				}
+				osEvents, err := osComponent.Events(c.rootCtx, time.Now().Add(-3*24*time.Hour))
+				if err != nil {
+					log.Logger.Errorw("failed to get os states", "error", err)
```

Comment:
I'm curious: was the continue missing because you didn't want to include it, or was it an oversight?

Reply:
wanted to skip it,  if we continue when os.Events fail we might incorrectly update state